### DISCUSSION TITLE
Using ImageSharp as a replacement for SkiaSharp for drawing to a GLFW window?

### DISCUSSION THREAD

**Discussion by bmitc** [SixLabors/ImageSharp.Drawing#233] [Ideas]
Hello! I am currently using SkiaSharp to draw to an OpenGL context provided by a GLFW window. When GLFW creates a window, it also creates an OpenGL context that is tied to the window. This context is then passed to SkiaSharp to draw to. An example F# test program:

```fsharp
let mutable width, height = 500, 500
let mutable framebufferWidth, framebufferHeight = width, height

let init = glfwInit() |> printfn "Initialized?: %A"

let window = glfwCreateWindow(width, height, "Test Window", NULL_MONITOR, NULL_WINDOW)

glfwMakeContextCurrent(window)

let grGlInterface = GRGlInterface.Create(getProcAddress)

if not (grGlInterface.Validate())
then raise (System.Exception("Invalid GRGlInterface"))

let grContext = GRContext.CreateGl(grGlInterface)

let draw(window, width, height) =
    glfwPollEvents()
    grContext.ResetContext()
    let grGlFramebufferInfo = new GRGlFramebufferInfo(0u, uint32(GL_RGBA8))
    let grBackendRenderTarget = new GRBackendRenderTarget(width, height, 1, 0, grGlFramebufferInfo)
    let surface = SKSurface.Create(grContext, grBackendRenderTarget, GRSurfaceOrigin.BottomLeft, SKColorType.Rgba8888)
    let canvas = surface.Canvas
    canvas.Clear(SKColors.Cyan)
    let red = new SKPaint(Color = SKColor(byte(255), byte(0), byte(0), byte(255)))
    canvas.DrawCircle(float32(width)/2.0f, float32(height)/2.0f, float32(100), red)
    canvas.Flush()
    glfwSwapBuffers(window)

while glfwWindowShouldClose window <> 1 do
    glfwGetFramebufferSize(window, &framebufferWidth, &framebufferHeight)
    draw(window, framebufferWidth, framebufferHeight)
    
glfwDestroyWindow(window)
terminate()
```

Can ImageSharp be used in a similar manner where I could use GLFW to create windows and use ImageSharp to draw to the window? Alternatives like using a Vulkan device instead of OpenGL are valid since I believe GLFW supports this, although I am not familiar with Vulkan yet.

I am having a lot of trouble with SkiaSharp and am investigating potential alternatives. The end goal is to use GLFW plus SkiaSharp or ImageSharp or whatever other 2D graphics library to build a GUI system for user interfaces, animations, creative coding, etc.

**Comment by antonfirsov** [comment]
Transferred this to the Drawing repostiory.

Note that ImageSharp and ImageSharp.Drawing do not have GPU acceleration, it's centered around in-memory image objects, and all rendering is done by the CPU. You cannot bind our images to window handles. The best thing you can do is to [wrap existing memory](https://docs.sixlabors.com/articles/imagesharp/memorymanagement.html#wrapping-existing-buffers-as-imagetpixel) as `Image<T>` if you somehow have a pointer to a pixel buffer that is going to be drawn, or do double buffering yourself with [`CopyPixelDataTo()`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Image-1.html#SixLabors_ImageSharp_Image_1_CopyPixelDataTo_).

**Reply by bmitc** [reply]
Thank you for the response! It sounds like the answer is that I can't reasonably replace SkiaSharp with ImageSharp for drawing to a window.

Can you elaborate on the difference between ImageSharp and ImageSharp.Drawing? I am unable to understand the differences between them from the READMEs. It seems like ImageSharp.Drawing just adds some additional high level behavior over ImageSharp?

My gut feeling is that calling ImageSharp a 2D graphics library is a bit of a misnomer. It is more an image processing library since it can't actually draw to a screen.

**Reply by antonfirsov** [reply]
> Can you elaborate on the difference between ImageSharp and ImageSharp.Drawing?

The ImageSharp is the a core image processing library, ImageSharp.Drawing adds drawing capabilities on top of that.

>  It is more an image processing library since it can't actually draw to a screen.

Personally, I tend to agree, especially when it comes to the core library, since I understand our definitions can be misleading for people coming from gamedev or similiar areas. On the other hand, the whole Fonts + ImageSharp + Drawing package is capable for *much* more than what many people would associate with "just image processing", and we need to communicate that in our branding. The core library's readme is a sort-of landing page.

**Reply by bmitc** [reply]
I see. Thanks for the insight! It's a tough battle given the overloaded meaning of "2D graphics".

> the whole Fonts + ImageSharp + Drawing package is capable for much more than what many people would associate with "just image processing"

Do you have any good examples of that? While ImageSharp doesn't currently fit my use case, I'd love to see examples of this! (Not doubting it. On the contrary, just wanting to learn of its capability.)

**Reply by antonfirsov** [reply]
An example is rendering non-latin (arabic, hagul ...) fonts aka. Font Shaping, which is an entire separate complex domain. AFAIK SkiaSharp does worse job than us out of the box, thanks to the advanced font shaper that is now being implemented in [SixLabors.Fonts](https://github.com/SixLabors/Fonts).

**Reply by antonfirsov** [reply]
> overloaded meaning of "2D graphics"

Also note that "image processing" is an overloaded term too. Some may associate it with machine learning stuff like image classification, or feature detection which is totally out of scope for us :)

**Reply by bmitc** [reply]
Haha. That's true regarding image processing.

Thanks for pointing to that example. Fonts are a bit out of my league at the moment, but I'm going to take a look to learn more.

My main goal right now is a cross-platform, desktop-only windowing and GUI platform for GUIs and "real-time" graphics (like Processing, vvvv, etc.). My issue is that all the GUI frameworks like MAUI, Uno, Avalonia, etc. are seriously overly complex, highly C#-centric, and desktop apps are usually an afterthought for those platforms.

For desktop windowing, GLFW is actually quite nice, and Skia and thus SkiaSharp ostensibly should support the graphics rendering, but I'm having trouble at the moment with SkiaSharp. And Skia is so complex, embarking on my own bindings, which I was easily able to do for GLFW, would be a big undertaking.

**Reply by antonfirsov** [reply]
If you like simple procedural drawing APIs and thin bindings, have you considered stuff like SDL or [bgfx](https://github.com/bkaradzic/bgfx)? (I have no experience with them in .NET) There might be many more out there, worth a research probably.
